### PR TITLE
Trigger upload on profiler shutdown.

### LINKED ISSUE #62: Profiler should trigger Upload on shutdown
Instead of uploading only by a time-based trigger, the profiler could launch the upload daemon again on shutdown. The upload daemon would then have to pass a message to any running daemon process to trigger an upload (after e.g. waiting a few seconds for safety).

This way, we could change the interval-based uploads to e.g. once per hour and on startup.

### PR DESCRIPTION
**Pull request by salsolatragus**
Fixes #62.

- [ ] Tests written
- [ ] Documentation updated

I prototyped a command forwarding that should allow us to control the `UploadDaemon` from the profiler. I'm unsure whether this solution covers all use cases we have and there's some decision I left open (TODOs in the code), because I lack information to decide what's the right approach. I would appreciate if we could have a talk about this, then it's probably straightforward to complete this.

### CODE REVIEW

**Comment by salsolatragus** [comment]
@karottenreibe I've rebased this on your Yaml config PR and implemented it according to our discussion last week. Actually made things much simpler. Rfr.

File: UploadDaemon/UploadDaemon.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        private readonly FileSystem fileSystem;
+
+        /// <summary>
+        /// Main entry point. Expects a single argument: the path to a directory that contains the trace files to upload.
+        /// </summary>
+        public static void Main(string[] args)
+        {
+            if (IsAlreadyRunning())
+            {
+                Console.WriteLine("Another instance is already running.");
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: please also mention that an upload has been triggered

File: UploadDaemon/UploadDaemon.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            if (config.DisableSslValidation)
+            {
+                HttpClientUtils.DisableSslValidation();
+            }
+
+            uploadTask = new UploadTask(config, fileSystem, new UploadFactory());
+        }
+
+        /// <summary>
+        /// Runs the timer action once synchronously.
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: you renamed the timer action. please update the comment

File: UploadDaemon/UploadDaemon.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        private void ScheduleRegularUploads()
+        {
+            Timer timer = new Timer();
+            timer.Elapsed += (sender, args) => RunOnce();
+            timer.Interval = TimerIntervalInMilliseconds;
+            timer.Enabled = true;
+        }
+
+        /// <summary>
+        /// Waits for notifications from subsequent executions of the Daemon. Terminates on
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: where is that handling for the cancel key? do we really need that?

File: UploadDaemon/UploadDaemon.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        private void WaitForNotifications()
+        {
+            while (true) // wait for indefinitely many commands
+            {
+                using (var pipeServerStream = new NamedPipeServerStream(DaemonControlPipeName, PipeDirection.In, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous))
+                {
+                    pipeServerStream.WaitForConnection();
+                    using (var streamReader = new StreamReader(pipeServerStream))
+                    {
+                        var command = streamReader.ReadLine();
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: do we have to trim or strip whitespace (i.e. the trailing \r\n) here?

File: UploadDaemon/UploadDaemon.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                        ExecuteCommand(command);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Executes a command forwarded to this UploadDaemon process.
+        /// </summary>
+        /// <param name="command"></param>
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: please remove useless param tag

File: UploadDaemon/UploadDaemon.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        private void ExecuteCommand(string command)
+        {
+            switch (command)
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: why do we need this switch statement if we always do the same thing? remove? this looks very confusing. I'd prefer a line comment that there is currently only one command and thus we don't need to check the command. And then you can inline this method actually

File: UploadDaemon/UploadDaemon.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        /// <summary>
+        /// Forwards a command to the existing UploadDaemon process.
+        /// </summary>
+        private static void NotifyRunningDaemon()
+        {
+            using (var pipeClientStream = new NamedPipeClientStream(".", DaemonControlPipeName, PipeDirection.Out, PipeOptions.Asynchronous))
+            {
+                pipeClientStream.Connect();
+
+                using (var w = new StreamWriter(pipeClientStream))
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: please don't use one-letter variable names

File: UploadDaemon/UploadDaemon.cs (hunk 3 above)
Comment (salsolatragus) [reply UploadDaemon/UploadDaemon.cs]: I removed this already, after we agreed we don't need it on the phone. Updated the comment now as well. Good catch!

File: UploadDaemon/UploadDaemon.cs (hunk 4 above)
Comment (salsolatragus) [reply UploadDaemon/UploadDaemon.cs]: Nope, [`ReadLine()` does that for us](https://docs.microsoft.com/en-us/dotnet/api/system.io.streamreader.readline?redirectedfrom=MSDN&view=netframework-4.7.2#remarks).

**Comment by salsolatragus** [comment]
rfr again

File: UploadDaemon/UploadDaemon.cs [hunk 8]
```diff
@@ -42,7 +41,7 @@ public static void Main(string[] args)
         {
             if (IsAlreadyRunning())
             {
-                Console.WriteLine("Another instance is already running.");
+                logger.Debug("Another instance is already running. Sending notification to trigger upload.");
```
Comment (karottenreibe) [inline UploadDaemon/UploadDaemon.cs]: this was written to the console on purpose: it's for the case where the user runs the executable from the command line. please change back to writeline. Logging with debug level will never be seen by anyone ;-)

**Pull request merged**